stdlib/src/blackbox_fallbacks/keccak256.rs

```diff
@@ -0,0 +1,267 @@
+//! Keccak256 fallback function.
+use super::{
+    sha256::pad,
+    uint8::UInt8,
+    utils::{byte_decomposition, round_to_nearest_byte},
+    UInt64,
+};
+use acir::{
+    circuit::Opcode,
+    native_types::{Expression, Witness},
+    FieldElement,
+};
+
+const BITS: usize = 256;
+const WORD_SIZE: usize = 8;
+const BLOCK_SIZE: usize = (1600 - BITS * 2) / WORD_SIZE;
+const ROUND_CONSTANTS: [u64; 24] = [
+    1,
+    0x8082,
+    0x800000000000808a,
+    0x8000000080008000,
+    0x808b,
+    0x80000001,
+    0x8000000080008081,
+    0x8000000000008009,
+    0x8a,
+    0x88,
+    0x80008009,
+    0x8000000a,
+    0x8000808b,
+    0x800000000000008b,
+    0x8000000000008089,
+    0x8000000000008003,
+    0x8000000000008002,
+    0x8000000000000080,
+    0x800a,
+    0x800000008000000a,
+    0x8000000080008081,
+    0x8000000000008080,
+    0x80000001,
+    0x8000000080008008,
+];
+const RHO: [u32; 24] =
+    [1, 3, 6, 10, 15, 21, 28, 36, 45, 55, 2, 14, 27, 41, 56, 8, 25, 43, 62, 18, 39, 61, 20, 44];
+const PI: [usize; 24] =
+    [10, 7, 11, 17, 18, 3, 5, 16, 8, 21, 24, 4, 15, 23, 19, 13, 12, 2, 20, 14, 22, 9, 6, 1];
+
+pub fn keccak256(
+    inputs: Vec<(Expression, u32)>,
+    outputs: Vec<Witness>,
+    mut num_witness: u32,
+) -> (u32, Vec<Opcode>) {
+    let mut new_gates = Vec::new();
+    let mut new_inputs = Vec::new();
+
+    // Decompose the input field elements into bytes and collect the resulting witnesses.
+    for (witness, num_bits) in inputs {
+        let num_bytes = round_to_nearest_byte(num_bits);
+        let (extra_gates, inputs, updated_witness_counter) =
+            byte_decomposition(witness, num_bytes, num_witness);
+        new_gates.extend(extra_gates);
+        new_inputs.extend(inputs);
+        num_witness = updated_witness_counter;
+    }
+
+    let (result, num_witness, extra_gates) = create_keccak_constraint(new_inputs, num_witness);
+    new_gates.extend(extra_gates);
+
+    // constrain the outputs to be the same as the result of the circuit
+    for i in 0..outputs.len() {
+        let mut expr = Expression::from(outputs[i]);
+        expr.push_addition_term(-FieldElement::one(), result[i]);
+        new_gates.push(Opcode::Arithmetic(expr));
+    }
+    (num_witness, new_gates)
+}
+
+fn create_keccak_constraint(
+    input: Vec<Witness>,
+    num_witness: u32,
+) -> (Vec<Witness>, u32, Vec<Opcode>) {
+    let mut new_gates = Vec::new();
+    let num_blocks = input.len() / BLOCK_SIZE + 1;
+
+    // pad keccak
+    let (input, extra_gates, mut num_witness) = pad_keccak(input, num_blocks, num_witness);
+    new_gates.extend(extra_gates);
+
+    // prepare state
+    let mut state = Vec::new();
+    for _ in 0..200 {
```

Comment:
```suggestion
    for _ in 0..STATE_NUM_BYTES {
```

Not wedded to the name but I think we should factor out some of the magic constants to be named.